src/rust/engine/nailgun/src/lib.rs

```diff
@@ -111,53 +111,89 @@ impl Server {
     config: nails::Config<N>,
     should_exit: oneshot::Receiver<()>,
-    state: watch::Sender<ServerState>,
+    exited: oneshot::Sender<Result<(), String>>,
     listener: TcpListener,
   ) {
     let exit_result = Self::accept_loop(executor, config, should_exit, listener).await;
     info!("Server exiting with {:?}", exit_result);
-    let _ = state.broadcast(ServerState::Exited(exit_result));
+    let _ = exited.send(exit_result);
   }
 
   async fn accept_loop<N: Nail>(
     executor: Executor,
     config: nails::Config<N>,
-    _should_exit: oneshot::Receiver<()>,
+    mut should_exit: oneshot::Receiver<()>,
     mut listener: TcpListener,
   ) -> Result<(), String> {
-    loop {
-      let tcp_stream = match listener.accept().await {
-        Ok((tcp_stream, _addr)) => tcp_stream,
-        Err(e) => {
+    // While connections are ongoing, they acquire `read`; before shutting down, the server
+    // acquires `write`.
+    let ongoing_connections = Arc::new(RwLock::new(()));
+
+    let result = loop {
+      let tcp_stream = match future::select(listener.accept().boxed(), should_exit).await {
+        future::Either::Left((Ok((tcp_stream, _addr)), s_e)) => {
+          // Got a connection.
+          should_exit = s_e;
+          tcp_stream
+        }
+        future::Either::Left((Err(e), _)) => {
           break Err(format!("Server failed to accept connections: {}", e));
         }
+        future::Either::Right((_, _)) => {
+          break Ok(());
+        }
       };
 
       debug!("Accepted connection: {:?}", tcp_stream);
-      let _join = executor.spawn(nails::server_handle_connection(config.clone(), tcp_stream));
-    }
+
+      // There is a slightly delicate dance here: we wait for a connection to have acquired the
+      // ongoing connections lock before proceeding to the next iteration of the loop. This
+      // prevents us from observing an empty lock and exiting before the connection has actually
+      // acquired it. Unfortunately we cannot acquire the lock in this thread and then send the
+      // guard to the other thread due to its lifetime bounds.
+      let connection_started = Arc::new(Notify::new());
+      let _join = executor.spawn({
+        let config = config.clone();
+        let connection_started = connection_started.clone();
+        let ongoing_connections = ongoing_connections.clone();
+        async move {
+          let ongoing_connection_guard = ongoing_connections.read().await;
+          connection_started.notify();
+          let result = nails::server_handle_connection(config.clone(), tcp_stream).await;
+          std::mem::drop(ongoing_connection_guard);
+          result
+        }
+      });
+      connection_started.notified().await;
+    };
+
+    // Before exiting, acquire write access on the ongoing_connections lock to prove that all
+    // connections have completed.
+    debug!("Server waiting for connections to complete...");
+    let _ = ongoing_connections.write().await;
```

Comment:
Ideally, all ongoing connection futures would be asynchronously cancelled here before the hard await. Is that possible?

Reply:
So, the gist of this change is that we actually want to wait for them to have fully completed. That might not be the right case for all consumers of this kind of server, but it's what we test for in the new test as well.

Reply:
Hrm. This sounds like it solves one problem but leads to another unless I'm missing something. If I CTRL-C a long-running goal I just started through the daemon, does that signal propagate to the nailgun handler so that it exits ~promptly?

Reply:
Yea, that would be an unclean/immediate shutdown. This is for clean/orderly shutdown when we've noticed things changing under us.